text/0000-cargo-asymmetric-tokens.md

```diff
@@ -0,0 +1,183 @@
+With a credential provider, the secret material can even stay on a hardware token.
+Furthermore, an asymmetric token can only be used for the intended action, and only for a short time window. The opportunity for replay is smaller, and can be tightened by the registry to meet its threat model.
+After the time window has expired, the data sent over the network can be made public, without risking the private material. A registry can keep or publish an audit log of asymmetric tokens without risk of them being reused, in case a security auditor would like to look for abnormal or unusual behavior.
+
+Different registries will have different users in mind and have different use cases. Therefore, they will need to have different behaviors. So, there are many decisions a registry has to make that this RFC has no opinion on. Some examples:
+
+- Bootstrapping trust: how does the registry decide to trust a new user?
+- Key generation: where and how the key pair is made?
+- Key rotation: how often do existing users need to make a new key pair?
+- Revocation: how does the registry decide to stop trusting an existing key pair?
+
+In order for crates.io to support asymmetric tokens these questions will need to be answered for crates.io. If and how crates.io will implement compatibility with these new tokens will be left for a follow-up discussion/RFC.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Private registries that require authentication use asymmetric cryptography as a more secure way for cargo to log in. Each registry works a little different, but the most common workflow is:
+1. Generate a key pair. (For many registries, you can generate the key pair using the cargo command `cargo login --generate-keypair`, which will print the public key for use in the next step.)
+2. Go to the registries log in page, upload your public key and get the user ID for that key pair.
+3. On the command line run `cargo login --registry=name --private-key-path=path userId`
+
+There are credential processes for using key pairs stored on hardware tokens. Check crates.io to see if there's one available for your hardware. Each one is a little different, but the general workflow is:
+1. `cargo install credential-process-for-your-hardware-token`
+2. run `cargo credential-process-for-your-hardware-token setup registryURL` to get your public key.
+3. Go to the registries log in page, upload your public key and get your user ID.
+4. Edit `credentials.toml` to have a `credential-process` field as described by `credential-process-for-your-hardware-token` docs. (The credential process command may help do this for you.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In [`config.toml`](https://doc.rust-lang.org/cargo/reference/config.html) and `credentials.toml` files there is a field called `private-key-path`, which provides a path to a `PKCS#12` formatted file containing a private key.
```

Comment:
I think storing the key directly in credentials.toml makes most sense, in base64 format or similar. Most users won't benefit from having it in a separate file. Instead, there is the burden of having *two* files that need to be protected, and it also means more complexity. Custom use cases can use a credentials process which is superior I think.

Reply:
I would suggest using PKCS#8 instead of PKCS#12. The latter is a container format for both PKCS#8 private keys (which are themselves asymmetric keypairs) and X.509 certificates, and X.509 certificates don't seem relevant to this particular application.

One nice thing about using PKCS#8 as opposed to something like credentials.toml is support for password-protected encrypted private keys, which can be used to protect these keys at rest. PKCS#8v2 supports deriving an encryption key using PBKDF2 or scrypt and using that to encrypt the original private key using AES.

Reply:
You can use `PKCS#8` or whatever other format, but ideally you don't store things into *yet another* file. That annoys users. toml supports multiline strings, so you can even have PEM encoding if you want.

Reply:
I was already convinced that cargo should only support one way of getting the key, and that all other ways of storing keys should require a credential process. I am now convinced that cargo should accept the (encoded) key directly as opposed to a path to the key.

The argument that convinced me is that most CI/CD systems allow you to store secrets in environment variables, but not particularly in files. If cargo accepts the key directly, then we already have an environment variable that will go straight from CIs secrets system to cargo. The same way it does for the existing secret tokens.

I did not have time to edit this into the RFC today, but I will do so and my next opportunity.